Maintenance app and force publish course view

mushtaqak opened the pull request:
**Problem**

Getting DevOps to run management commands takes quite a bit of time to escalate the critical issues quickly. We often wait for DevOps to execute the commands and then observe the results, so a lot of important time is lost.

**Solution**
Have a maintenance dashboard for the escalation team that let them interact with advance views for executing the management commands ( which would be replaced by the views eventually).

More on Maintenance App [here](https://openedx.atlassian.net/wiki/display/SUST/Maintenance+Board)

Here is the basic layout of the `Maintenance App`.

<img width="1220" alt="screen shot 2016-05-06 at 6 02 45 pm" src="https://cloud.githubusercontent.com/assets/6991154/15073798/fc31bb16-13b4-11e6-93c9-f8cf1abec7c3.png">

Home page adds the the shortcut to maintenace app via a user-account-dropdown

<img width="1219" alt="screen shot 2016-05-06 at 6 02 32 pm" src="https://cloud.githubusercontent.com/assets/6991154/15073772/dcc5c9ca-13b4-11e6-9f6e-a9565eb8a460.png">

`studio_url/maintenance` url shows the list of the views available

<img width="1222" alt="screen shot 2016-05-06 at 6 03 09 pm" src="https://cloud.githubusercontent.com/assets/6991154/15073829/1c0b06a4-13b5-11e6-9aff-6f04d0e17e07.png">


`studio_url/maintenance/force_publish_course` url : shows force publish course view



And the rest of the pages would follow the same pattern, Here is the list of all the commands that we may want.

`studio_url/maintenance/show_orphans`
`studio_url/maintenance/delete_orphans`
`studio_url/maintenance/export_course`
`studio_url/maintenance/import_course`
`studio_url/maintenance/delete_course`

<img width="1202" alt="screen shot 2016-04-27 at 8 21 16 pm" src="https://cloud.githubusercontent.com/assets/6991154/15073876/54343280-13b5-11e6-96a9-ad545fbc9f67.png">


the`/maintenance` would look like this when these views are implemented.

[SUST-19](https://openedx.atlassian.net/browse/SUST-19) , [SUST-46](https://openedx.atlassian.net/browse/SUST-46)

adampalay commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,126 @@
+def show_orphans(request):
+    """Render show orphans view """
+    # Only global staff (PMs) are able to activate/deactivate certificate configuration
+    if not GlobalStaff().has_user(request.user):
```
you can use https://github.com/edx/edx-platform/blob/master/lms/djangoapps/instructor/views/api.py#L246 instead of this conditional logic

mushtaqak replied:
Thanks for this decorator. I was  going to find this after I had finished the skeleton of the app. But you saved my time :)

adampalay commented on cms/templates/maintenance/show_orphans.html:
```diff
@@ -0,0 +1,18 @@
+        View Orphans : Not yet Implemented
+    </section>
+</div>
+</%block>
```
nit: add newline

adampalay commented on cms/templates/maintenance/index.html:
```diff
@@ -0,0 +1,18 @@
+    </ul>
+</section>
+</div>
+</%block>
```
nit: add newline

adampalay commented on cms/templates/maintenance/import_course.html:
```diff
@@ -0,0 +1,18 @@
+        Import Course: Not yet Implemented
+    </section>
+</div>
+</%block>
```
nit: add newline

adampalay commented on cms/templates/maintenance/force_publish_course.html:
```diff
@@ -0,0 +1,18 @@
+        Force Publish Course : Not yet Implemented
+    </section>
+</div>
+</%block>
```
nit: add newline

adampalay commented on cms/templates/maintenance/export_course.html:
```diff
@@ -0,0 +1,18 @@
+        Export Course : Not yet Implemented
+    </section>
+</div>
+</%block>
```
nit: add newline

adampalay commented on cms/templates/maintenance/delete_orphans.html:
```diff
@@ -0,0 +1,18 @@
+        Deleted Orphans : Not yet Implemented
+    </section>
+</div>
+</%block>
```
nit: add newline

adampalay commented on cms/templates/maintenance/delete_course.html:
```diff
@@ -0,0 +1,18 @@
+        Delete Course : Not yet Implemented
+    </section>
+</div>
+</%block>
```
nit: add newline

adampalay commented on cms/djangoapps/maintenance/urls.py:
```diff
@@ -0,0 +1,17 @@
+    url(r'^export_course/?$', views.export_course, name="export_course"),
+    url(r'^import_course/?$', views.import_course, name="import_course"),
+    url(r'^delete_course/?$', views.delete_course, name="delete_course"),
+    url(r'^force_publish_course/?$', views.force_publish_course, name="force_publish_course"),
```
rather than have stub pages, can you implement just one of these (like `force_publish_course` or `export_course`)? Then you can delete the management commands associated with them.

mushtaqak replied:
I just created the skeleton today. I will implement one command and move to another after that. Can we create some SUST ticket for that ? I would like to go with `print_orphans` / `force_publish_course` command initially. What do you say ?

adampalay replied:
+1

mushtaqak commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,164 @@
+"""
+Index view for the maintenance app.
```
Need to change this line ... ugh

mushtaqak commented:
@adampalay You may want to look again :)

adampalay commented on common/djangoapps/util/views.py:
```diff
@@ -44,6 +47,21 @@ def inner(request, *args, **kwargs):
     return inner
 
 
+def require_global_staff(func):
```
shouldn't we just move the original into "common" so both lms and csm can use it?

adampalay replied:
oh whoops, that's what you did :)

adampalay commented on cms/templates/maintenance/base.html:
```diff
@@ -0,0 +1,51 @@
+            <div class="wrapper wrapper-nav-sub">
+              <div class="nav-sub">
+                <ul>
+                  <li class="nav-item nav-account-dashboard">
```
provide translator notes

adampalay commented on cms/templates/maintenance/base.html:
```diff
@@ -0,0 +1,51 @@
+      </h1>
+      <h2 class="info-course">
+        <a class="course-link" href="${reverse('maintenance:maintenance')}">
+          <span>Maintenance Dashboard</span>
```
should be i18n

adampalay commented on cms/templates/maintenance/_show_orphans.html:
```diff
@@ -0,0 +1,60 @@
+<%page expression_filter="h"/>
```
why are we implementing this if there's already a studio view that does this?

adampalay commented on cms/templates/maintenance/_import_course.html:
```diff
@@ -0,0 +1,3 @@
+<div class="wrap-instructor-info studio-view">
+    Import Course : Not yet Implemented
```
I don't think we should have these stubs

adampalay commented on cms/templates/maintenance/index.html:
```diff
@@ -0,0 +1,18 @@
+from django.utils.translation import ugettext as _
+from django.core.urlresolvers import reverse
+%>
+<%block name="title">Maintenance Dashboard</%block>
```
i18n

adampalay commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,192 @@
+        return render_to_response('maintenance/container.html', context)
+
+
+class DeleteOrphansView(View):
```
Let's not implement these stubs

mushtaqak commented:
[Maintenance app sandbox](https://studio-maintenance.sandbox.edx.org/maintenance) added.

@adampalay @awais786 Please review :)

awaisdar001 commented on cms/djangoapps/contentstore/management/commands/force_publish.py:
```diff
@@ -1,71 +0,0 @@
-                else:
-                    print "Dry run. Course '{0}' is already in published state.".format(course_key)
-        else:
-            raise CommandError("The owning modulestore does not support this command.")
```
Are we deprecating force publish mgm command once we have a GUI version? I mean, is this explicit?

mushtaqak replied:
Yes, we plan to remove the management commands as we implement them in the web views.

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,209 @@
+            self.context['msg'] = COURSE_KEY_ERROR_MESSAGES['empty_course_key']
+            return course_usage_key
+        try:
+            if branch == "published":
```
I think validity of passed `course_key` should checked prior to adding branch information.

mushtaqak replied:
Yes, but once `course_usage_key` is created, it can not be changed. It is immutable.

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,209 @@
+            if branch == "published":
+                course_key += "+branch@published-branch"
+
+            course_usage_key = CourseKey.from_string(course_key)
```
What I we have an invalid course key pattern?

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,209 @@
+                'error': False,
+                'msg': ''
+            })
+        except InvalidKeyError:
```
Oh I see the logic here :)

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,209 @@
+        course_id = request.POST.get('course-id')
+        is_dry_run = bool(request.POST.get('dry-run'))
+
+        course_usage_key = self.validate_course_key(course_id)
```
Can we test if `course_usage_key` is not None, CC: Row#86

mushtaqak replied:
We can surely  test, but if it is None, then we already have set `context['error']` which would do the same work.

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,209 @@
+        if self.context['error']:
+            return self.render_response()
+
+        owning_store = modulestore()._get_modulestore_for_courselike(course_usage_key)  # pylint: disable=protected-access
```
`owning_store`?

mushtaqak replied:
`owning_store` is the modulestore of the course. Like if we have passed `course-v1:edx+DemoX+Demo` it would be split modeulestore if course key is `edx/XSS/101` it would be mongo modulestore.

We are checking this because, force publish does not work for mongo courses.

awaisdar001 replied:
should be `source_store` and `destination_store`?

awaisdar001 commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,228 @@
+        """Create a user and log in. """
+        super(MaintenanceViewTestCase, self).setUp()
+        self.user = AdminFactory()
+        success = self.client.login(username=self.user.username, password='test')
```
`success`?

mushtaqak replied:
To check if user was able to log in.

awaisdar001 replied:
Sorry if I am causing a pain or var names, but i think `login_success` makes more sense to me. :)

awaisdar001 commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,228 @@
+    @ddt.data(*(
+        (url_name, role, has_access)
+        for (url_name, (role, has_access))
+        in itertools.product((
```
Please explain this, Not so readable.

mushtaqak replied:
I am checking this for each URL, we have different roles which would either have access or not.

awaisdar001 commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,228 @@
+        ))
+    ))
+    @ddt.unpack
+    def test_access(self, url_name, role, has_access):
```
I think we should have separate test for `user_has_access` and `has_not_access`. I think we should split it up.

awaisdar001 commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,228 @@
+        self.verify_error_message({'course-id': course_key}, error_msg)
+
+    def verify_bad_course_keys(self):
+        self.verify_invalid_course_key('', COURSE_KEY_ERROR_MESSAGES['empty_course_key'])
```
that should be injected via dd.data to the test instead of having a method

awaisdar001 commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,228 @@
+        self.validate_success_from_response(response, _('Now published and draft branch have same version'))
+
+        # now course is forcefully published, we should get already published course.
+        self.verify_error_message(
```
again in a single test we are testing multiple scenarios.

awaisdar001 commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,228 @@
+        versions = get_course_versions(unicode(self.course.id))
+
+        # verify that draft and publish point to different versions
+        self.assertNotEqual(versions['draft-branch'], versions['published-branch'])
```
Name suggests to me that method is going to force publish, but we are also checking versioning logic here. Can we extract it  and let the method do what it sounds like? Just a suggestion, I found it unrelated to check version in `force_publish_course` method.

awaisdar001 commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,228 @@
+        self.validate_success_from_response(response, _('Following course versions would be changed'))
+
+        # verify that course still has changes as we just dry ran force publish course.
+        self.assertTrue(self.store.has_changes(self.store.get_item(self.course.location)))
```
Can we check the version did not changed also during dry run?

awaisdar001 commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,228 @@
+    @ddt.unpack
+    def test_access(self, url_name, role, has_access):
+        user = UserFactory(username='test', email="test@example.com", password="test")
+        success = self.client.login(username=user.username, password='test')
```
`login_success` i think :)

mushtaqak commented:
@adampalay waiting for your review too

mushtaqak commented:
@awaisdar001 addressed all your comments. 

@adampalay Ping

adampalay commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,243 @@
+            {
+                'course-id': course_key
+            },
+            _(error_message)
```
no reason to use ugettext in a test

adampalay commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,243 @@
+        # force publish course
+        data = {
+            'course-id': unicode(self.course.id),
+            'dry-run': is_dry_run
```
shouldn't this be a `bool`?

adampalay commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,243 @@
+        """
+        Test complete flow of force publish as dry run.
+        """
+        response = self.force_publish_course(is_dry_run='on')
```
shouldn't `is_dry_run` be `True` or `False`

adampalay commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,243 @@
+        Test complete flow of force publish as dry run.
+        """
+        response = self.force_publish_course(is_dry_run='on')
+        self.validate_success_from_response(response, _('Following course versions would be changed'))
```
no need to use `uggettext`

adampalay commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,243 @@
+        versions = get_course_versions(unicode(self.course.id))
+
+        response = self.force_publish_course()
+        self.validate_success_from_response(response, _('Now published and draft branch have same version'))
```
no need for `ugettext` here

adampalay commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,243 @@
+        """
+        Test complete flow of force publish.
+        """
+        versions = get_course_versions(unicode(self.course.id))
```
maybe rename to `old_versions`?

adampalay commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,217 @@
+            'msg': ''
+        }
+
+    def render_response(self):
```
why is this a separate method?

adampalay commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,217 @@
+
+        try:
+            if branch == "published":
+                course_key += "+branch@published-branch"
```
instead of manipulating the string, use methods already defined on the course key:
```python
In [9]: ck = CourseKey.from_string("course-v1:a+b+c")

In [10]: ck.for_branch?
Signature: ck.for_branch(branch)
Docstring: Return a new CourseLocator for another branch of the same course (also version agnostic)
File:      ~/venvs/edxapp/local/lib/python2.7/site-packages/opaque_keys/edx/locator.py
Type:      instancemethod

In [11]: unicode(ck.for_branch('published-branch'))
Out[11]: u'course-v1:a+b+c+branch@published-branch'
```

adampalay commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,217 @@
+
+        current_versions = get_course_versions(course_id)
+
+        # if publish and draft were different
```
this should read: `if publish and draft were NOT different`

adampalay commented on cms/templates/base.html:
```diff
@@ -65,7 +65,10 @@
     <div class="wrapper wrapper-view" dir="${static.dir_rtl()}">
-        <% online_help_token = self.online_help_token() if hasattr(self, 'online_help_token') else None %>
+        <%
+        #from nose.tools import set_trace;set_trace()
```
you should remove this :)

adampalay replied:
it looks like this file shouldn't be changed at all?

adampalay commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,81 @@
+        % if form_data['is_dry_run']:
+        <article>
+            <div class="dry-run">
+                ${Text(_("You have {strong_start}dry run{strong_end} the course. Following course versions would be changed :")).format(
```
"You have done a dry run of force publishing the course. Nothing has changed. Had you run it, the following course versions would have been changed"

adampalay commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,81 @@
+                )}
+            </div>
+            <div class="main-output">
+                ${_("Published branch version {published} changed to draft branch version {draft}").format(
```
`The published branch version, {published}, was reset to the draft branch version, {draft}`

adampalay commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,81 @@
+        % elif updated_versions:
+        <article>
+            <div class="main-output">
+                ${Text(_("{strong_start}Success!{strong_end} Published the course {course_id} forcefully.")).format(
```
"Force published", not "forcefully"

adampalay commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,243 @@
+
+
+@ddt.ddt
+class MaintenanceViewTestCase(ModuleStoreTestCase):
```
use [`SharedModuleStoreTestCase`](http://swampcastle.org/2015/07/28/writing-faster-modulestore-tests.html) since most of these tests don't actually change state. It'll speed it up a lot!

adampalay replied:
@mushtaqak , turns out that decorator that allowed you to modify courseware was removed, so no need to use `SharedModuleStoreTestCase`

adampalay commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,217 @@
+
+log = logging.getLogger(__name__)
+
+MAINTENANCE_COMMANDS = {
```
I don't know why this feels weird to me. I feel like these attributes should be defined on the `ForcePublishCourseView`. The index view can maintain a list of the views that should show up under it.

mushtaqak replied:
Actually, Index view lists down all the commands along with their name and description and needs URL to forward to the specific view. 

This also helps in getting the list of all the URLs in the maintenance app which is used by many tests.

adampalay replied:
You can still get all that if you define these on the index view.
in sudo-python:
```python
def get_urls():
    urls = []
    for view in IndexView.views:
        urls.append(view.url)
```

or something like that.

These attributes should be defined on each class-based view

mushtaqak replied:
OK I can move it to IndexView but I would still need those attributes in IndexView as they are used to print description, name and url of the view.

mushtaqak replied:
This kind of approach is also used in LMS Support app here https://github.com/edx/edx-platform/blob/a680919e8c9331b21e9a27d2abd7e3985a8b1a0e/lms/djangoapps/support/views/index.py#L11-L11

adampalay commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,217 @@
+
+class MaintenanceIndexView(View):
+    """
+    View for maintenance dashboard, used by the escalation team.
```
Index view for maintenance dashboard...

Also, you might want to explain what it's used for

adampalay commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,217 @@
+    template = 'maintenance/container.html'
+
+    def __init__(self, command=None):
+        self.context = {
```
I don't like the idea of having context be this very mutable thing that different functions just kind of change. Like the fact that `validate_course_key` mutates content isn't really clear. That function should either return the course_key or raise an error that we then handle and change or create the correct context appropriately. Right now, it's hard to reason about what's in context, and I can imagine this being difficult to debug.

adampalay replied:
ping :)

mushtaqak replied:
I am doing this because every view would have context variable with the exact same data initially. So it is better if we set it in BaseView's init method.

adampalay commented:
@mushtaqak , first pass complete. Eventually we'll want an accessibility review, but maybe after we finish an escalation review

mushtaqak commented:
@adampalay addressed comments, please review :)

adampalay commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,243 @@
+
+
+@ddt.ddt
+class MaintenanceViewTestCase(SharedModuleStoreTestCase):
```
@mushtaqak , I'm not sure you're using the using the `SharedModuleStoreTestCase` correctly. The idea is that you set up your course in a `setUpClass` method that's a classmethod (so that it's only set up once per test class). Here's Dave's original blog post: http://swampcastle.org/2015/07/28/writing-faster-modulestore-tests.html

I'm happy to walk you through this tomorrow morning if that would be helpful!

mushtaqak replied:
I will look into this though :)

mushtaqak replied:
@adampalay Isn't publishing the course action of modifying it ? So why do we want to use `SharedModuleStoreTestCase` which should only be used if individual tests do not modify the course. As described in [SharedModuleStoreTestCase](https://github.com/edx/edx-platform/pull/9070/files#diff-480f335443b1a6487aeb0282e7d6cd31R222) class which states :

> Use this class if your tests set up one or a small number of courses that individual tests do not modify. If your tests modify contents in the ModuleStore, you should use ModuleStoreTestCase instead.

mushtaqak replied:
@adampalay I couldn't find the decorator that your saying yesterday. 

I could only find `override_settings` decorator but that also only affects settings.

adampalay commented:
@mushtaqak , I still have questions about `self.context` and the `MAINTENANCE_COMMANDS` dict

mushtaqak commented:
@adampalay I have replaced `SharedModuleStoreTestCase` with `ModuleStoreTestCase`  since I could not find the decorator that can be used on any test that modifies the course. Also, I renamed the `MAINTENANCE_COMMANDS` dict to `MAINTENANCE_VIEWS` which is more clear.

Please see what else can we improve more ?

@awaisdar001 ping on review :)

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,231 @@
+    Returns all URLs for maintenance app.
+    """
+    url_list = []
+    for key, val in MAINTENANCE_VIEWS.items():  # pylint: disable=unused-variable
```
We need to improve that. 
User list comprehension.

adampalay replied:
1. instead of using `val`, use a variable name like `view_metadata`
2. I agree with @awaisdar001 , a list comprehension should be nicer here
3. if you're just using a value, do `MAINTENANCE_VIEWS.values()` or `MAINTENANCE_VIEWS.itervalues()`

awaisdar001 commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,243 @@
+        """
+        # validate non split error message
+        course = CourseFactory.create(default_store=ModuleStoreEnum.Type.mongo)
+        self.verify_error_message(
```
This need to be updated.

awaisdar001 commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,243 @@
+        Argument:
+
+            is_dry_run - default True, that means by default the view does dry run.
+        """
```
Can we use pep8 Docstring Conventions? I am open. 
https://www.python.org/dev/peps/pep-0257/

awaisdar001 commented:
just couple of comments :)

adampalay commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,79 @@
+                            <label for="dry-run" class="label" data-tooltip="${_('Trial or practice run the task')}">${_("Dry Run")}</label>
+                            % if form_data['is_dry_run']:
+                            <% is_checked = "checked" %>
+                            % else :
```
nit: no space between `else` and `:`

adampalay commented:
couple nits, otherwise looks good to me. we'll need an a11y review

awaisdar001 commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,243 @@
+
+    def test_force_publish_dry_run(self):
+        """
+        Test complete flow of force publish as dry run.
```
Can we describe what is `complete flow`. Please mention explicitly.

awaisdar001 commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,243 @@
+
+    def test_force_publish(self):
+        """
+        Test complete flow of force publish.
```
Same here.

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,231 @@
+
+class ForcePublishCourseView(MaintenanceBaseView):
+    """
+    View for force publishing state of the course, used by the escalation team.
```
`escalation team` or global staff? :)

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,231 @@
+        source_store = modulestore()._get_modulestore_for_courselike(course_usage_key)  # pylint: disable=protected-access
+        if not hasattr(source_store, 'force_publish_course'):
+            self.context['msg'] = _("Force publish course does not support old mongo style courses.")
+            logging.info(
```
`logging` -> `logger`
+
logger.warning will be much more suitable.

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,231 @@
+
+        source_store = modulestore()._get_modulestore_for_courselike(course_usage_key)  # pylint: disable=protected-access
+        if not hasattr(source_store, 'force_publish_course'):
+            self.context['msg'] = _("Force publish course does not support old mongo style courses.")
```
`Force publishing is not supported with mongo courses.`

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,231 @@
+        )
+        if not updated_versions:
+            self.context['msg'] = _("Could not publish course.")
+            logging.info(
```
warning.

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,231 @@
+            course_usage_key, request.user, commit=True
+        )
+        if not updated_versions:
+            self.context['msg'] = _("Could not publish course.")
```
More readable message? Why can not publish?

mushtaqak replied:
Now that I have gone deeper into this, I think, this message and check won't be necessary. Because, the only way `updated_versions` is None is when we have faulty course locator which we have already taken care in the very start of the post method.

awaisdar001 replied:
Sounds like we can get rid of it?

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,231 @@
+        # if publish and draft are NOT different
+        if current_versions['published-branch'] == current_versions['draft-branch']:
+            self.context['msg'] = _("Course is already in published state.")
+            logging.info(
```
I am not sure `info` is the right thing here, `error` or `warning` should be considered.

awaisdar001 commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,243 @@
+        response = self.client.get(url)
+
+        # Expect a redirect to the login page
+        redirect_url = "{login_url}?next={original_url}".format(
```
Use single quotes here and everywhere else besides docstrings.

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,231 @@
+# This dict maintains all the views that will be used Maintenance app.
+MAINTENANCE_VIEWS = {
+    "force_publish_course": {
+        "url": reverse_lazy("maintenance:force_publish_course"),
```
Use single quotes

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,231 @@
+            published_prev=current_versions['published-branch'],
+            published_new=updated_versions['published-branch']
+        )
+        logging.info(
```
`.debug`

awaisdar001 commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,231 @@
+        try:
+            course_usage_key = self.validate_course_key(course_id)
+        except InvalidKeyError:
+            self.context['error'] = True
```
Do we need to also log here?
`logger.exception`?

mushtaqak replied:
I don't think so, this would be over-logging I guess. It is just a user input error.

mushtaqak commented:
@adampalay @awaisdar001 addressed the comments.

@cptvitamin Can we have a a11y review ? 

Here is the [sandbox](https://studio-maintenance.sandbox.edx.org/).

You can access [maintenance app](https://studio-maintenance.sandbox.edx.org/maintenance).

adampalay commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,242 @@
+        self.user = AdminFactory()
+        login_success = self.client.login(username=self.user.username, password='test')
+        self.assertTrue(login_success)
+        self.course = CourseFactory.create(default_store=ModuleStoreEnum.Type.split)
```
Don't make the course here... course creation can slow down tests, so only use it when you need it!

mushtaqak commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -2295,10 +2296,9 @@ def list_report_downloads(_request, course_id):
     response_payload = {
         'downloads': [
-            dict(name=name, url=url, link='<a href="{}">{}</a>'.format(url, name))
+            dict(name=name, url=url, link=HTML('<a href="{}">{}</a>').format(url, name))
```
@robrap What is the best way to fix [`python-wrap-html`](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/safe_templates.html?highlight=xss#python-wrap-html) violation.

I have fixed this one using HTML wrapper, but this seems failing the test which can't find the JSON.

robrap replied:
@mushtaqak: This fix generally looks fine.  Unfortunately, the test failure needs additional digging to determine what is actually going wrong.  Let me know when you hone in on why this is breaking.

efischer19 replied:
I have run into this exact same issue, here's some output that may be helpful.
```
ERROR: test_list_report_downloads (lms.djangoapps.instructor.tests.test_api.TestInstructorAPILevelsDataDump)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/mock.py", line 1618, in _inner
    return f(*args, **kw)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/instructor/tests/test_api.py", line 2849, in test_list_report_downloads
    res_json = json.loads(response.content)
  File "/usr/lib/python2.7/json/__init__.py", line 338, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 366, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 384, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
-------------------- >> begin captured stdout << ---------------------
{
  "downloads": [
    {
      "url": "https://1.mock.url", 
      "link": &#34;<a href=\&#34;https://1.mock.url\&#34;>mock_file_name_1</a>&#34;, 
      "name": "mock_file_name_1"
    }, 
    {
      "url": "https://2.mock.url", 
      "link": &#34;<a href=\&#34;https://2.mock.url\&#34;>mock_file_name_2</a>&#34;, 
      "name": "mock_file_name_2"
    }
  ]
}
```

efischer19 replied:
The problem is coming from this line: https://github.com/edx/edx-platform/blob/master/common/djangoapps/util/json_request.py#L69

```
> /edx/app/edxapp/edx-platform/common/djangoapps/util/json_request.py(69)__init__()
-> content = json.dumps(resp_obj, cls=encoder, indent=2, ensure_ascii=False)
(Pdb) n
> /edx/app/edxapp/edx-platform/common/djangoapps/util/json_request.py(70)__init__()
-> kwargs.setdefault("content_type", "application/json")
(Pdb) content
u'{\n  "downloads": [\n    {\n      "url": "https://1.mock.url", \n      "link": &#34;<a href=\\&#34;https://1.mock.url\\&#34;>mock_file_name_1</a>&#34;, \n      "name": "mock_file_name_1"\n    }, \n    {\n      "url": "https://2.mock.url", \n      "link": &#34;<a href=\\&#34;https://2.mock.url\\&#34;>mock_file_name_2</a>&#34;, \n      "name": "mock_file_name_2"\n    }\n  ]\n}'
(Pdb) resp_obj
{'downloads': [{'url': 'https://1.mock.url', 'link': Markup(u'<a href="https://1.mock.url">mock_file_name_1</a>'), 'name': 'mock_file_name_1'}, {'url': 'https://2.mock.url', 'link': Markup(u'<a href="https://2.mock.url">mock_file_name_2</a>'), 'name': 'mock_file_name_2'}]}
```

mushtaqak commented:
jenkins run bokchoy

mushtaqak commented:
@cptvitamin ping.

cptvitamin commented on cms/templates/maintenance/base.html:
```diff
@@ -0,0 +1,21 @@
+<div class="wrapper-content wrapper">
+    <div class="maintenance-header">
+        <h2>
+            <a href="${reverse('maintenance:maintenance')}">
```
Is there a reason why this appears as a link while you are on the maintenance dashboard?

mushtaqak replied:
I have added this in the block, so that this will be available in all child templates. All will have the link so this would too have the link so as to make templates consistent.

cptvitamin commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,80 @@
+    <form class="form-create" method="post">
+        <input type="hidden" name="csrfmiddlewaretoken" value="${csrf_token}"/>
+        <div class="wrapper-form">
+            <fieldset>
```
I'm not sure we want to use a ```<fieldset>``` here. The fieldset element requires that it's first child be a ```<label>``` element.

mushtaqak replied:
I was using the the form from Sign Up page that's why I used this one.

cptvitamin replied:
But that form uses a fieldset with a legend. This one doesn't.  The sign up form adds the sr class to the legend to move it offscreen, which is fine.  Something similar here would work.  It would be helpful if the legend contained text that described what the form will do once submitted.

cptvitamin commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,80 @@
+        <input type="hidden" name="csrfmiddlewaretoken" value="${csrf_token}"/>
+        <div class="wrapper-form">
+            <fieldset>
+                <ol class="list-input">
```
Is there a particular reason we are using the ordered list element here?  This doesn't seem to be a numbered list of items, especially since there are no visual numbers showing.  We should only be using lists to markup actual lists.

mushtaqak replied:
I was actually using the same form used on the [Sign Up page](https://github.com/edx/edx-platform/blob/master/cms/templates/register.html#L30)

cptvitamin replied:
I understand that, but its not right there either.  lists should only be used to markup lists.

cptvitamin commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,80 @@
+                        <label for="course-id">${_('Course ID')}</label>
+                        <input id="course-id" type="text" name="course-id" aria-required="true"
+                               value="${form_data['course_id']}"
+                               placeholder="${_('e.g. edX+DemoX+Demo_Course')}"/>
```
I would like to see this placeholder text either moved to another location that is visible on screen all the time, or duplicated there.  Placeholder text disappears once the person starts typing.  In this particular example, the user would probably like to have access to the placeholder value once they start typing.  Can we put it below the text field in a ```<span>```?  The ```<input>``` would then require an aria-describedby attribute that references the IDref of the span.

mushtaqak replied:
If I use the `<span>` then should I still user placeholder ?

cptvitamin replied:
you can use both, but it seems unnecessary.  The span that is visible all the time is all that is needed.

cptvitamin commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,80 @@
+                <ol class="list-input">
+                    <li class="field text required" id="field-course-name">
+                        <label for="course-id">${_('Course ID')}</label>
+                        <input id="course-id" type="text" name="course-id" aria-required="true"
```
aria-required means that this is ONLY required for screen reader users. I doubt that is the intention. ```required``` makes it required for everyone.

mushtaqak replied:
Makes sense.

cptvitamin commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,80 @@
+                               value="${form_data['course_id']}"
+                               placeholder="${_('e.g. edX+DemoX+Demo_Course')}"/>
+                        % if error:
+                        <div class="error">${msg}</div>
```
To make this accessible, you should author it this way:
```
<!-- the error message container should be on the DOM at page load, even if empty. It should be referenced  on the input with aria-describedby -->
<input id="course-id" type="text" name="course-id" aria-describedby="cid-error" required value="${form_data['course_id']}"
  placeholder="${_('e.g. edX+DemoX+Demo_Course')}"/>
<!-- the error message container should have an aria-live=polite so that it reads the message to the user when updated-->
<div class="error" aria-live="polite" id="cid-error"> 
% if error: 
  <%  $msg %>
% endif
</div>
``` 

You should also move the focus to the form field (input) that has the first error.

mushtaqak replied:
makes sense.

cptvitamin commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,80 @@
+                    </li>
+                    <li class="field-radio">
+                        <div>
+                            <label for="dry-run" class="label" data-tooltip="${_('Trial or practice run the task')}">
```
this tooltip is only available to mouse users.  It needs to be accessible to keyboard and screen reader users too.  consider the example above (put it on screen and reference it via aria-describedby)

cptvitamin commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,80 @@
+            </fieldset>
+        </div>
+        <div class="actions">
+            <button type="submit" class="action action-primary" aria-disabled="true">${_('Force Publish Course')}
```
again, this button will not work for a screen reader user *only*.   If you want to disable it for everyone, then ```disabled``` works.

cptvitamin commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,80 @@
+        <div class="actions">
+            <button type="submit" class="action action-primary" aria-disabled="true">${_('Force Publish Course')}
+            </button>
+            <a href="${view['url']}" class="action action-secondary action-cancel"
```
Seems very odd to have one be a button and the other be a link.  Also, the data-tooltip needs to be accessible to everyone.  Recommend just making the visible text "Reset values"

mushtaqak replied:
well caught. I'll change to button :)

mushtaqak replied:
Just remembered that we needed a GET request that is why I used the anchor element.

cptvitamin commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,80 @@
+    % if current_versions:
+    <div class="result">
+        % if form_data['is_dry_run']:
+        <article>
```
Seems an inappropriate use of the ```<article>``` element:
https://www.w3.org/TR/html5/sections.html#article-or-section?

Should be a section or a div.

mushtaqak replied:
I will change to `div`

cptvitamin commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,80 @@
+            </div>
+        </article>
+        % elif updated_versions:
+        <article>
```
section or div

cptvitamin commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,80 @@
+    </div>
+    % elif not error:
+    <div class="wrapper-error">
+        <div class="error" role="alert">${msg}</div>
```
this div should be visible on the page at page load (otherwise the role=alert will not work).  Put the conditional inside this element.

mushtaqak replied:
makes sense.

cptvitamin commented on cms/templates/maintenance/index.html:
```diff
@@ -0,0 +1,17 @@
+%>
+<%block name="title">${_('Maintenance Dashboard')}</%block>
+<%block name="viewcontent">
+<section class="container maintenance-content">
```
If this is truly a ```<section>``` then it should have a heading as its first child that describes the theme of the content.  Perhaps you should make the two blocks that precede this children of the section element.  That would work.

mushtaqak replied:
So Should I add some paragraph describing Maintenace Dashboard or should I make this one `div` too ?

mushtaqak commented:
@cptvitamin Thanks for reviewing :) I will reply and address the comments.

mushtaqak commented:
@cptvitamin I have addressed and replied to your comments. Please see. Thanks :)

mushtaqak commented on cms/templates/widgets/header.html:
```diff
@@ -229,6 +230,11 @@ <h3 class="title"><span class="label"><span class="label-prefix sr">${_("Current
                   <li class="nav-item nav-account-dashboard">
                     <a href="/">${_("{studio_name} Home").format(studio_name=settings.STUDIO_SHORT_NAME)}</a>
                   </li>
+                  % if GlobalStaff().has_user(user):
```
@cptvitamin @adampalay I had added this dropdown option to user-nav. Is this okay ?

mushtaqak commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,87 @@
+        <div class="actions">
+            <button type="submit" class="action action-primary">${_('Force Publish Course')}
+            </button>
+            <a href="${view['url']}" class="action action-secondary action-cancel"
```
@adampalay and @awaisdar001 should we add some javascript for this page ? If that it is in the scope then we would need multiple changes.

1. Ajax calls on POST
2. Ajax response output.
3. Proper Javascript validations.
4. May be backbone views ?

We initially thought we don't need to go into deep, but now to cover a11y, do we need these things done ? 

@cptvitamin FYI

adampalay replied:
@mushtaqak , I don't see why we need javascript, unless it's more accessible that way?

mushtaqak replied:
@adampalay Actually, there are multiple purposes for adding javascript. Just like in the last comment you asked :

> Please disable buttons after you click them to prevent users from clicking multiple times before the requests go through

So, for that I would need Javascript and I have two ways.

One way being adding javascript in `js_extra` block, just like :

```
<%block name="js_extra">
   <script type="text/javascript">
   (function() {
     $("#button_id").click(function(e) {
       // operations to perform
       e.preventDefault();
     })
</script>
</%block>
```

Which is not a good way, we would need to do this for every view, it is not a clean solution.

Second way is that we create backbone view/models for these and do all things from those. 

Second approach is better but it will take good  amount of time but code will look more better and clean.

I think we can create a separate SUST ticket to accomplish that. What do you say ?

adampalay replied:
What about just making a separate javascript file without making the backbone views? We already have a helper function for this: https://github.com/edx/edx-platform/blob/master/common/static/common/js/components/utils/view_utils.js#L118-L124

mushtaqak replied:
OK - I let me see, how can we accomplish this in better way with less time :) 

I will update tomorrow :)

mushtaqak commented:
@cptvitamin I have addressed all the comments. Please review :) Thanks.

I have created the latest [maintenance app sandbox](https://studio-maintenance.sandbox.edx.org/maintenance) 

@adampalay , @awaisdar001 Can you review again. Thanks.

awaisdar001 commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,267 @@
+
+    def test_non_split_course(self):
+        """
+        Test that we get a error message on non-split courses.
```
As we are specifically testing `mongo` modulestore, I think this should be `... error message on mongo courses.`

adampalay commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,267 @@
+        Reverse the setup.
+        """
+        self.client.logout()
+        ModuleStoreTestCase.tearDown(self)
```
nit: should this be `super(MaintenanceViewTestCase, self).setUp()`

adampalay commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,267 @@
+        # force publish course view
+        data = {
+            'course-id': unicode(course.id),
+            'dry-run': 'on' if is_dry_run else ''
```
should this just be `'dry-run': is_dry_run`?

mushtaqak replied:
No, we need else to complete the statement otherwise there is a compiler error.

adampalay replied:
I mean why is this a string instead of a bool?

mushtaqak replied:
actually, this is a checkbox which sends 'on' value when we send form POST request. I wanted to make them consistent.

adampalay replied:
Oh, interesting. I had no idea that's what happened when you submitted a POST. Do you mind adding a comment to that effect?

adampalay commented on cms/templates/maintenance/_force_publish_course.html:
```diff
@@ -0,0 +1,88 @@
+                        <label for="course-id">${_('Course ID')}</label>
+                        <input id="course-id" type="text" name="course-id" aria-describedby="course-id-desc" required
+                               value="${form_data['course_id']}"/>
+                        <div id="course-id-desc" class="tip tip-stacked">${_('e.g. edX+DemoX+Demo_Course')}</div>
```
nit: you'll want something like `course-v1:edX+DemoX+Demo_Course`

adampalay commented on common/djangoapps/util/json_request.py:
```diff
@@ -66,7 +66,7 @@ def __init__(self, resp_obj=None, status=None, encoder=EDXJSONEncoder,
             content = serialize('json', resp_obj)
         else:
-            content = json.dumps(resp_obj, cls=encoder, indent=2, ensure_ascii=False)
+            content = json.dumps(resp_obj, cls=encoder, indent=2, ensure_ascii=True)
```
@mushtaqak , why did you change this? Is it necessary? Will this affect other things?

robrap replied:
@adampalay: You can see the discussion on this here: https://github.com/edx/edx-platform/pull/12301/files/6de52f6496a4fe01b58a1c854bee59f1e2315a6a#r64092353

adampalay replied:
ah, thanks @robrap  :)

adampalay commented:
@mushtaqak , just some small comments. once you address them, 👍 . This is really good work, thanks :)

adampalay commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,220 @@
+        'url': reverse_lazy('maintenance:force_publish_course'),
+        'name': _('Force Publish Course'),
+        'slug': 'force_publish_course',
+        'description': _('Force publish course.'),
```
Can you add a more descriptive description? Maybe something like, "Sometimes the draft and published branches of a course can get out of sync. This command resets the published branch of a course to point to the draft branch, effectively force publishing the course."

adampalay commented:
@mushtaqak , I actually have one final suggestion: Please disable buttons after you click them to prevent users from clicking multiple times before the requests go through

mushtaqak commented:
@adampalay while build checks are running, can you please have a look at this again. I have addressed your and @awaisdar001 comments and changed the index view CSS to accommodate larger text descriptions.

mushtaqak commented on cms/envs/common.py:
```diff
@@ -819,6 +819,9 @@
     'openedx.core.djangoapps.coursetalk',  # not used in cms (yet), but tests run
     'xblock_config',
 
+    # Maintenance tools
```
@adampalay @awaisdar001 I also added maintenance app to INSTALLED_APPS which I missed earlier.

cptvitamin commented:
@mushtaqak this LGTM 👍  There are a number of contrast issues on this page but we should be able to resolve those later when we do more concentrated contrast work on Studio.

adampalay commented on cms/djangoapps/maintenance/views.py:
```diff
@@ -0,0 +1,223 @@
+        'url': reverse_lazy('maintenance:force_publish_course'),
+        'name': _('Force Publish Course'),
+        'slug': 'force_publish_course',
+        'description': _('Sometimes the draft and published branches of a course can get out of sync.'
```
You need spaces between each line in a multiline string if you want those spaces to appear. Also reformat:

```
'description': _(
    'Sometimes the draft and published branches of a course can get out of sync. '
    'This command resets the published branch of a course to point to the draft branch, '
    'effectively force publishing the course.'
),
```

adampalay commented on cms/djangoapps/maintenance/tests.py:
```diff
@@ -0,0 +1,268 @@
+        self.assertTrue(login_success)
+
+        response = self.client.get(url)
+        self.assertContains(response, 'Must be edX staff to perform this action.', status_code=403)
```
this test will fail if `settings.PLATFORM_NAME` isn't `edX`. I don't know if that's a problem, tho

adampalay commented:
@mushtaqak , a couple small comments. I'd like to see the "force publish" button grayed out when you click it, but if you want to do that in a separate PR, be my guest. 👍 once my comments are addressed

mushtaqak commented:
@jenkins run python

mushtaqak commented:
jenkins run a11y

mushtaqak commented:
@adampalay may you please have a look ?

adampalay commented:
@mushtaqak 👍 once tests pass

awaisdar001 commented:
👍 After rebase please.

mushtaqak commented:
jenkins run python

mushtaqak commented:
jenkins run bokchoy

mushtaqak commented:
jenkins run python

mushtaqak commented:
@adampalay I am getting `cms.djangoapps.contentstore.views.tests.test_certificates.CertificatesDetailHandlerTestCase.*` test failures, I had rebased the branch too, any idea ?

mushtaqak commented:
jenkins run python

mushtaqak commented:
These test are passing locally.

awaisdar001 commented:
@mushtaqak If the tests keep failing can you ask @Ben?

adampalay commented:
Yeah, these pass for me locally too. @mushtaqak , can you check in the testeng hipchat room?

mushtaqak commented:
jenkins run python

mushtaqak commented:
jenkins run python

cpennington commented:
I think the easiest thing to do is to add the `UrlResetMixin` to the `CertificatesBaseTestCase`. That should fix your problem.

mushtaqak commented:
jenkins run python

mushtaqak commented:
@cpennington Still getting the same url issue on `cms.djangoapps.contentstore.views.tests.test_header_menu.TestHeaderMenu.test_header_menu_with_web_certs_enabled`. I have also added the `UrlResetMixin` to `TestHeaderMenu` but this one fails still. Any ideas ?

mushtaqak commented:
jenkins run python

mushtaqak commented:
jenkins run bokchoy

mushtaqak commented:
@adampalay Once all tests pass, am I good to merge `Maintenance App?

adampalay commented on cms/djangoapps/contentstore/views/tests/test_certificates.py:
```diff
@@ -197,7 +197,12 @@ def test_certificate_data_validation(self):
 @ddt.ddt
 @override_settings(FEATURES=FEATURES_WITH_CERTS_ENABLED)
-class CertificatesListHandlerTestCase(EventTestMixin, CourseTestCase, CertificatesBaseTestCase, HelperMethods):
+class CertificatesListHandlerTestCase(EventTestMixin,
```
breaking long lines! http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python_guidelines.html#breaking-long-lines

adampalay commented on cms/djangoapps/contentstore/views/tests/test_certificates.py:
```diff
@@ -420,7 +426,12 @@ def test_assign_unique_identifier_to_certificates(self):
 @ddt.ddt
 @override_settings(FEATURES=FEATURES_WITH_CERTS_ENABLED)
-class CertificatesDetailHandlerTestCase(EventTestMixin, CourseTestCase, CertificatesBaseTestCase, HelperMethods):
+class CertificatesDetailHandlerTestCase(EventTestMixin,
```
http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python_guidelines.html#breaking-long-lines

adampalay commented:
@mushtaqak one style nit, then a squash and rebase, then 👍

mushtaqak commented:
jenkins run lettuce

adampalay commented:
🎆
